crates/db_schema/src/impls/oauth_provider.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{
+  newtypes::OAuthProviderId,
+  schema::oauth_provider,
+  source::oauth_provider::{
+    OAuthProvider,
+    OAuthProviderInsertForm,
+    OAuthProviderUpdateForm,
+    UnsafeOAuthProvider,
+  },
+  traits::Crud,
+  utils::{get_conn, DbPool},
+};
+use diesel::{dsl::insert_into, result::Error, QueryDsl};
+use diesel_async::RunQueryDsl;
+
+#[async_trait]
+impl Crud for UnsafeOAuthProvider {
+  type InsertForm = OAuthProviderInsertForm;
+  type UpdateForm = OAuthProviderUpdateForm;
+  type IdType = OAuthProviderId;
+
+  async fn create(pool: &mut DbPool<'_>, form: &Self::InsertForm) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    insert_into(oauth_provider::table)
+      .values(form)
+      .get_result::<Self>(conn)
+      .await
+  }
+
+  async fn update(
+    pool: &mut DbPool<'_>,
+    oauth_provider_id: OAuthProviderId,
+    form: &Self::UpdateForm,
+  ) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    diesel::update(oauth_provider::table.find(oauth_provider_id))
+      .set(form)
+      .get_result::<Self>(conn)
+      .await
+  }
+}
+
+impl UnsafeOAuthProvider {
+  pub async fn get_all(pool: &mut DbPool<'_>) -> Result<Vec<Self>, Error> {
+    let conn = &mut get_conn(pool).await?;
+    let oauth_providers = oauth_provider::table
+      .order(oauth_provider::id)
+      .select(oauth_provider::all_columns)
```

Comment:
Isn't the only difference that `client_secret` is included in the unsafe?

In that case you might not need any of the extra types and functions below, since you already have `serde_skip` on that field. 

Check out the `site` table, which just has `pub private_key: Option<SensitiveString>,` , and doesn't need a separate `SiteSafe` / `SiteUnsafe` version.

@Nutomic might be able to give better guidance than I can on this.

Reply:
We could also just add serde skip to all of those also. 

I'm good with either way I spose, but I'd like to hear @Nutomic 's thoughts on this.

Reply:
I would definitely like to get rid of these duplicate struct and implementations because it makes maintenance more difficult. Though I agree that we shouldnt include unnecessary data in `GetSiteResponse` to avoid bloated data. To keep it simple I would consider always sending these fields for admin accounts, and skipping them for normal users. I would also move the fields into a separate struct (and table?) so this can be done more easily.

Reply:
I don't fully understand why that's necessary, you could just:

- Eliminate the duplicate struct, and use serde_skip for all the necessary fields.
- Make oauth_providers in `GetSiteResponse` optional, and only include it for admins.

Reply:
Sweet